### ISSUE TITLE
Make Layer.affected_tables private

### ISSUE THREAD

**Issue by javitonino** [CartoDB/cartodb#10939]
- `Layer.affected_tables` calculates the affected tables from the query, analyses, etc.
- `Layer.register_table_dependencies` takes the previous output and writes it to the DB (layers_user_table)
- `Layer.user_table` returns the cached version of the affected tables from the DB

I think we should make `affected_tables` private to enforce used of the cached version. Right now, `affected_tables` is being used in handful of places around the code (e.g: vizjson3 presenter, .carto exporter) but it makes no sense to use this instead of the caching version (assuming the cached version is correct, which we should strive to be the case).

cc @CartoDB/builder-backend opinions?

**Comment by gfiorav** [comment]
Go for it!

**Comment by juanignaciosl** [comment]
👍  as long as current usages don't explicitly need an uncached one (and even in that case the call should probably cache it, so a `force_uncached` option might be better).

In addition, `register_table_dependencies` is needed as public?

**Comment by javitonino** [comment]
Oh, there are more uses than I thought, via the visualization relator (old model). In fact, there are more uncached used than there are cached ones. So if we go for this, I would consider running a rake to rebuild all caches.

Seeing as most uses now are uncached, I would like to do some benchmarking of cached vs uncached first.

> In addition, `register_table_dependencies` is needed as public?
It's called from the LayersController, but most likely that call can be removed, and this made private.

**Comment by ivanmalagon** [comment]
Found a somehow related bug while doing acceptance.

Steps taken:

1. Create a map linked to a given dataset.
2. Create another map link to the same dataset.
3. Go back to the fist map and delete the shared dataset.
4. Go to dataset dashboard and click on delete that dataset.
5. A warning shows up referring to the both maps, when only one is actually linked.

Move back this issue to development and block it while @javitonino research on it and creates another ticket for this case. Can't accept this one until that is resolved because from the UI point of view the result is not the expected (even if the bug it's in another code place).

**Comment by javitonino** [comment]
Unblocked :)

**Comment by ivanmalagon** [comment]
Acceptance done in ded02.

Issue with dataset in two maps, deleting the dataset in one of them and warning not being refreshed not happening anymore.

Tested scenarios:
- Create empty map.
- Import dataset from file.
- Import synced file from URL.
- Import datasets from .carto.
- Referring to dataset in SQL.
- Analyses that can refer to another table (intersect second layer, join columns from 2nd layer)
- Same dataset in several maps.
- Combinations of deleting / updating datasets in several maps.

So far, so good. :+1:
